politeiawww/invoices.go

```
@@ -358,30 +408,284 @@ func (p *politeiawww) processInvoiceDetails(invDetails cms.InvoiceDetails,
+			err)
+	}
+
+	// Verify the SetUnvettedStatus challenge.
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, pdReply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	// Update the database with the metadata changes.
+	dbInvoice.Changes = append(dbInvoice.Changes, cmsdatabase.InvoiceChange{
+		Timestamp:      changes.Timestamp,
+		AdminPublicKey: changes.AdminPublicKey,
+		NewStatus:      changes.NewStatus,
+		Reason:         changes.Reason,
+	})
+	dbInvoice.StatusChangeReason = changes.Reason
+	dbInvoice.Status = changes.NewStatus
+
+	err = p.cmsDB.UpdateInvoice(dbInvoice)
+	if err != nil {
+		return nil, err
+	}
+	/*
+		p.fireEvent(EventTypeInvoiceStatusChange,
+			EventDataInvoiceStatusChange{
+				Invoice:   dbInvoice,
+				AdminUser: user,
+			},
+		)
+	*/
+	// Return the reply.
+	sisr := cms.SetInvoiceStatusReply{
+		Invoice: *convertDatabaseInvoiceToInvoiceRecord(*dbInvoice),
+	}
+	return &sisr, nil
+}
+
+func validateStatusTransition(
+	oldStatus cms.InvoiceStatusT,
+	newStatus cms.InvoiceStatusT,
+	reason string,
+) error {
+	validStatuses, ok := validStatusTransitions[oldStatus]
+	if !ok {
+		log.Errorf("status not supported: %v", oldStatus)
+		return www.UserError{
+			ErrorCode: www.ErrorStatusInvalidInvoiceStatusTransition,
+		}
+	}
+
+	if !statusInSlice(validStatuses, newStatus) {
+		return www.UserError{
+			ErrorCode: www.ErrorStatusInvalidInvoiceStatusTransition,
+		}
+	}
+
+	if newStatus == cms.InvoiceStatusRejected && reason == "" {
+		return www.UserError{
+			ErrorCode: www.ErrorStatusReasonNotProvided,
+		}
+	}
+
+	return nil
+}
+
+func statusInSlice(arr []cms.InvoiceStatusT, status cms.InvoiceStatusT) bool {
+	for _, s := range arr {
+		if status == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+// processEditInvoice attempts to edit a proposal on politeiad.
+func (p *politeiawww) processEditInvoice(ei cms.EditInvoice, u *user.User) (*cms.EditInvoiceReply, error) {
```

Submitting an invoice edit without any changes gets accepted and creates a new version.  This should return a user error.

I think you should just be able to compute the file digests and compare them against the existing file digests.

---

Ah good point, hrm ok I'll do that